zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,491 @@
+// Copyright 2023 The Fuchsia Authors
+//
+// Licensed under a BSD-style license <LICENSE-BSD>, Apache License, Version 2.0
+// <LICENSE-APACHE or https://www.apache.org/licenses/LICENSE-2.0>, or the MIT
+// license <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your option.
+// This file may not be copied, modified, or distributed except according to
+// those terms.
+
+use ::proc_macro2::TokenStream;
+use ::quote::quote;
+use ::syn::{DataEnum, Fields, Generics, Ident, Variant};
+use syn::Index;
+
+use crate::{EnumRepr, Trait};
+
+struct RawIdents {
+    r#enum: Ident,
+    discriminant: Ident,
+    fields: Ident,
+    variant_prefix: String,
+}
+
+impl RawIdents {
+    const DISCRIMINANT_VALUE_PREFIX: &'static str = "RAW_DISCRIMINANT_";
+
+    fn for_enum(ident: &Ident) -> Self {
+        Self {
+            r#enum: Ident::new(&format!("Raw{}Enum", ident), ident.span()),
+            discriminant: Ident::new(&format!("Raw{}Discriminant", ident), ident.span()),
+            fields: Ident::new(&format!("Raw{}Fields", ident), ident.span()),
+            variant_prefix: format!("Raw{}Variant", ident),
+        }
+    }
+
+    fn discriminant_const(&self, variant: &Ident) -> Ident {
+        Ident::new(&format!("{}{}", Self::DISCRIMINANT_VALUE_PREFIX, variant), variant.span())
+    }
+
+    fn variant_struct(&self, variant: &Ident) -> Ident {
+        Ident::new(&format!("{}{}", self.variant_prefix, variant), variant.span())
+    }
+}
+
+/// Generates an implementation of `is_bit_valid` for the given enum input.
+pub(crate) fn derive_is_bit_valid(
+    reprs: &[EnumRepr],
+    ident: &Ident,
+    generics: &Generics,
+    data: &DataEnum,
+) -> TokenStream {
+    let idents = RawIdents::for_enum(ident);
+
+    if reprs.contains(&EnumRepr::C) {
+        derive_is_bit_valid_c(data, generics, &idents)
+    } else {
+        derive_is_bit_valid_primitive(data, generics, discriminant_repr(reprs), &idents)
+    }
+}
+
+pub(crate) fn discriminant_repr(reprs: &[EnumRepr]) -> &EnumRepr {
+    if reprs.contains(&EnumRepr::C) {
+        reprs.iter().find(|&r| r != &EnumRepr::C).unwrap_or(&EnumRepr::C)
+    } else {
+        reprs.first().unwrap()
+    }
+}
```

Comment:
Add a comment explaining why we're guaranteed this won't miss anything? E.g., why `#[repr(u8, u16)]` is impossible (since we'd only return the `u8` part).

Reply:
Added lots of comments